google-cloud-pubsub/lib/google/cloud/pubsub/async_publisher/batch.rb

```diff
@@ -0,0 +1,293 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+require "monitor"
+require "google/cloud/pubsub/errors"
+
+module Google
+  module Cloud
+    module PubSub
+      class AsyncPublisher
+        ##
+        # @private
+        class Batch
+          include MonitorMixin
+
+          attr_reader :items, :ordering_key
+
+          def initialize publisher, ordering_key
+            @publisher = publisher
+            @ordering_key = ordering_key
+            @items = []
+            @queue = []
+            @default_message_bytes = publisher.topic_name.bytesize + 2
+            @total_message_bytes = @default_message_bytes
+            @publishing = false
+            @stopping = false
+            @canceled = false
+
+            # init MonitorMixin
+            super()
+          end
+
+          ##
+          # Adds a message and callback to the batch.
+          #
+          # The method will indicate how the message is added. It will either be
+          # added to the active list of items, it will be queued to be picked up
+          # once the active publishing job has been completed, or it will
+          # indicate that the batch is full and a publishing job should be
+          # created.
+          #
+          # @param [Google::Cloud::PubSub::V1::PubsubMessage] msg The message.
+          # @param [Proc, nil] callback The callback.
+          #
+          # @return [Symbol] The state of the batch.
+          #
+          #   * `:added` - Added to the active list of items to be published.
+          #   * `:queued` - Batch is publishing, and the messsage is queued.
+          #   * `:full` - Batch is full and ready to be published, and the
+          #     message is queued.
+          #
+          def add msg, callback
+            synchronize do
+              raise AsyncPublisherStopped if @stopping
+              raise OrderingKeyCanceled, @ordering_key if @canceled
+
+              if @publishing
+                queue_add msg, callback
+                :queued
+              elsif try_add msg, callback
```

Comment:
It looks like there is a scenario where messages could get out of order. If I call publish on a message that will exceed the size of a batch, then it will get added to the queue. If I then call publish on a message that does not exceed the batch size, it will get added to the batch. Now these messages will be delivered out of order.

Reply:
Messages will not be added out of order. When the publish is complete the batch will pull all available messages from the queue before allowing any message to be added to the batch directly. All this happens in the same lock.

Reply:
Okay, I see how this works. As soon as the batch is full, the `AsyncPublisher` will immediately call `publish_batches!` which marks _all_ batches as publishing (on the caller's thread, and in the `AsyncPublisher`'s lock, so that it is not possible for multiple calling threads to race.) This then causes subsequent adds to go directly to the queue until publishing has completed.

I agree this will work. I'm concerned about fragility, however, since the correctness of the batch class's state (i.e. ordering of messages across the items and the queue) depends on internal logic in the publisher class and the correct use of the publisher class's synchronization. I would prefer to see the batch class be thread-safe within itself, which you could accomplish by having this `add` method itself update any relevant state to indicate where future adds should go.

Reply:
This was discussed offline. It turns out that _all_ batches need to move into a publishing phase at the same time (as soon as one batch fills up), which makes this more complicated than can be consolidated into a single method on `Batch`. I still think there might be a way to make the logic more solid, but it will take some more thought. Since we're confident of "correctness" at this point, let's accept the code as is and think about refactoring later.